### PR TITLE
Feature 5162 - LimitIfTouched Orders

### PR DESCRIPTION
**Pull request by aarjaneiro**
#### Description

Currently there is no implementation of the LimitIfTouchedOrder.
This is an order type where a trigger price is defined along with a limit price for a limit order to be placed upon "touching" the trigger. Specifically, versus a stop limit order, the limit order is placed in the opposite direction with respect the the trigger/stop which is first touched.
e.g.,
For sell orders:
```
      ------------/--------------- Limit price
             /\/\/
         /\/
------/------------------------- Trigger touched => places limit order 
___/
```
#### Related Issue

#5162 
#### Motivation and Context

One can only synthetically create a LIT order but never can submit/interact with non-synthetic IBKR LIT orders
#### How Has This Been Tested?

Adds tests under relevant fill model test files.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [ x] My code follows the code style of this project.
- [ x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [ x] I have added tests to cover my changes. 
- [ x] All new and existing tests passed.
- [ x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs [hunk 1]
```diff
@@ -1737,6 +1738,12 @@ private IBApi.Order ConvertOrder(Order order, Contract contract, int ibOrderId)
                 ibOrder.LmtPrice = Convert.ToDouble(RoundPrice(stopLimitOrder.LimitPrice, minTick));
                 ibOrder.AuxPrice = Convert.ToDouble(RoundPrice(stopLimitOrder.StopPrice, minTick));
             }
+            else if (limitIfTouchedOrder != null)
+            {
+                var minTick = GetMinTick(contract, order.Symbol);
+                ibOrder.LmtPrice = Convert.ToDouble(RoundPrice(limitIfTouchedOrder.LimitPrice, minTick));
+                ibOrder.TriggerPrice = Convert.ToDouble(RoundPrice(limitIfTouchedOrder.TriggerPrice, minTick));
```
Comment (Martin-Molinero) [inline Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs:1745]: Cool! 👍

File: Common/Orders/Fills/EquityFillModel.cs [hunk 2]
```diff
@@ -101,6 +106,109 @@ public virtual Fill Fill(FillModelParameters parameters)
...
+        {
+            //Default order event to return.
+            var utcTime = asset.LocalTime.ConvertToUtc(asset.Exchange.TimeZone);
+            var fill = new OrderEvent(order, utcTime, OrderFee.Zero);
+
+            //If its cancelled don't need anymore checks:
+            if (order.Status == OrderStatus.Canceled) return fill;
+
+            // make sure the exchange is open/normal market hours before filling
+            if (!IsExchangeOpen(asset, false)) return fill;
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:133]: Should we allow extended market hours fills, like `Limit` orders do here? Same for the `FillModel`

File: Common/Orders/Fills/EquityFillModel.cs [hunk 3]
```diff
@@ -101,6 +106,109 @@ public virtual Fill Fill(FillModelParameters parameters)
...
+
+            // do not fill on stale data
+            if (endTimeUtc <= order.Time) return fill;
+
+            //Check if the limit if touched order was filled:
+            switch (order.Direction)
+            {
+                case OrderDirection.Buy:
+                    //-> 1.2 Buy: If Price below Trigger, Buy:
+                    if (low <= order.TriggerPrice || order.TriggerTouched)
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:176]: `low <= order.TriggerPrice` hm not sure about this, the trigger price can be above the current market price too and should only get triggered when reached by a trade? Should be covered by unit tests 👍 suggest testing with IB

File: Common/Orders/Fills/EquityFillModel.cs [hunk 4]
```diff
@@ -101,6 +106,109 @@ public virtual Fill Fill(FillModelParameters parameters)
...
+                    {
+                        order.TriggerTouched = true;
+
+                        if (low <= order.LimitPrice)
+                        {
+                            fill.Status = OrderStatus.Filled;
+                            fill.FillPrice = open >= order.LimitPrice? open + slip // Fill price conditional on open
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:183]: Here we are buying: `open >= order.LimitPrice? open + slip // Fill price conditional on open` the limit price can't fill worse than the limit price, it can fill better though: `Using a Limit if Touched order helps to ensure that, if the order does execute, the order will not execute at a price less favorable than the limit price.` at https://www1.interactivebrokers.com/en/index.php?f=592
Behavior should be covered by unit tests 👍

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Nice one! Leaving some comments and concerns 👍

File: Common/Securities/CashBuyingPowerModel.cs [hunk 5]
```diff
@@ -371,6 +371,10 @@ private static decimal GetOrderPrice(Security security, Order order)
                 case OrderType.StopLimit:
                     orderPrice = ((StopLimitOrder)order).LimitPrice;
                     break;
+                
+                case OrderType.LimitIfTouched:
+                    orderPrice = ((StopLimitOrder)order).LimitPrice;
```
[the remark this answers is not in the document]
Comment (aarjaneiro) [reply Common/Securities/CashBuyingPowerModel.cs:376]: Whoops!

File: Common/Orders/Fills/EquityFillModel.cs [hunk 6]
```diff
@@ -101,6 +106,109 @@ public virtual Fill Fill(FillModelParameters parameters)
...
+        /// </summary>
+        /// <param name="asset">Security asset we're filling</param>
+        /// <param name="order">Order packet to model</param>
+        /// <returns>Order fill information detailing the average price and quantity filled.</returns>
+        /// <remarks>
+        ///     There is no good way to model limit orders with OHLC because we never know whether the market has
+        ///     gapped past our fill price. We have to make the assumption of a fluid, high volume market.
+        ///
+        ///     With limit if touched, we can't be sure of the order of the H - L values for the limit fill. The assumption
+        ///     was made the limit fill will be done with closing price of the bar after the stop has been triggered.
```
[the remark this answers is not in the document]
Comment (aarjaneiro) [reply Common/Orders/Fills/EquityFillModel.cs:121]: As is relatively apparent, I took to modelling the structure of LIT based off of Stop Limit orders (as they are relatively similar and wherever SLO was, an LIT implementation would also need to have been introduced). Clearly a few artifacts remain (e.g., stop →trigger, in this case)

File: Common/Orders/LimitIfTouchedOrder.cs [hunk 7]
```diff
@@ -0,0 +1,140 @@
+using System;
+using QuantConnect.Interfaces;
+using QuantConnect.Securities;
+using static QuantConnect.StringExtensions;
+
+namespace QuantConnect.Orders
+{
+    /// <summary>
+    /// <para>
```
[the remark this answers is not in the document]
Comment (aarjaneiro) [reply Common/Orders/LimitIfTouchedOrder.cs:9]: I used paras mostly to control newlines but I reckon here it provides nothing

File: Algorithm.CSharp/MarginCallEventsAlgorithm.cs [hunk 8]
```diff
@@ -72,7 +72,7 @@ public override void OnMarginCall(List<SubmitOrderRequest> requests)
                 // liquidate an extra 10% each time we get a margin call to give us more padding
                 var newQuantity = (int)(Math.Sign(order.Quantity) * order.Quantity * 1.1m);
                 requests.Remove(order);
-                requests.Add(new SubmitOrderRequest(order.OrderType, order.SecurityType, order.Symbol, newQuantity, order.StopPrice, order.LimitPrice, Time, "OnMarginCall"));
+                requests.Add(new SubmitOrderRequest(order.OrderType, order.SecurityType, order.Symbol, newQuantity, order.StopPrice, order.LimitPrice, order.TriggerPrice, Time, "OnMarginCall"));
```
[the remark this answers is not in the document]
Comment (aarjaneiro) [reply Algorithm.CSharp/MarginCallEventsAlgorithm.cs:75]: Admittedly, I think SubmitOrderRequest could do with a few more constructors for each order type!

File: Common/Orders/Fills/FillModel.cs [hunk 9]
```diff
@@ -291,7 +296,75 @@ public virtual OrderEvent StopLimitFill(Security asset, StopLimitOrder order)
...
+            var slip = asset.SlippageModel.GetSlippageApproximation(asset, order);
+            switch (order.Direction)
+            {
+                case OrderDirection.Sell:
+                    if (prices.High > order.TriggerPrice || order.TriggerTouched)
+                    {
+                        order.TriggerTouched = true;
+
+                        //-> 1.1 Limit surpassed: Sell.
+                        if (prices.Low > order.LimitPrice)
```
[the remark this answers is not in the document]
Comment (aarjaneiro) [reply Common/Orders/Fills/FillModel.cs:335]: Was trying to avoid using Current for the reason it is avoided in equityfillmodels (but that doesn't apply here)

File: Common/Orders/Order.cs [hunk 10]
```diff
@@ -347,41 +347,94 @@ public static Order FromSerialized(SerializedOrder serializedOrder)
...
+                serializedOrder.LimitPrice ?? 0,
+                serializedOrder.StopPrice ?? 0,
+                serializedOrder.TriggerPrice).PriceCurrency = serializedOrder.PriceCurrency;
+            CreateOrder(serializedOrder.OrderId, serializedOrder.Type, symbol, serializedOrder.Quantity,
+                DateTime.SpecifyKind(createdTime, DateTimeKind.Utc),
+                serializedOrder.Tag,
+                new OrderProperties { TimeInForce = timeInForce },
+                serializedOrder.LimitPrice ?? 0,
+                serializedOrder.StopPrice ?? 0,
+                serializedOrder.TriggerPrice).Status = serializedOrder.Status;
```
[the remark this answers is not in the document]
Comment (aarjaneiro) [reply Common/Orders/Order.cs:395]: Odd, I think my rebase was polluted

File: Common/Orders/Fills/EquityFillModel.cs (hunk 3 above)
Comment (aarjaneiro) [reply Common/Orders/Fills/EquityFillModel.cs:176]: > suggest testing with IB
Will do!

File: Common/Orders/Fills/EquityFillModel.cs (hunk 4 above)
Comment (aarjaneiro) [reply Common/Orders/Fills/EquityFillModel.cs:183]: Good eye, should have been a `<=` !

File: Common/Orders/OrderEvent.cs [hunk 11]
```diff
@@ -241,6 +258,10 @@ public override string ToString()
             {
                 message += Invariant($" StopPrice: {StopPrice.Value.SmartRounding()}");
             }
+            if (TriggerPrice.HasValue)
+            {
+                message += Invariant($" StopPrice: {TriggerPrice.Value.SmartRounding()}");
```
Comment (Martin-Molinero) [inline Common/Orders/OrderEvent.cs:263]: `StopPrice:` message is wrong

Comment (aarjaneiro) [reply Common/Orders/OrderEvent.cs:263]: Good catch!

File: Common/Orders/OrderEvent.cs [hunk 12]
```diff
@@ -124,6 +125,22 @@ public decimal FillQuantity
                 }
             }
         }
+        
+        /// <summary>
+        /// The current trigger price
+        /// </summary>
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public decimal? TriggerPrice
```
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Common/Orders/OrderEvent.cs:133]: We recently added `ShortToString()` let's add it there to, sorry about that 😄

File: Common/Orders/Fills/FillModel.cs [hunk 13]
```diff
@@ -286,6 +291,88 @@ public virtual OrderEvent StopLimitFill(Security asset, StopLimitOrder order)
...
+            
+            //Get the range of prices in the last bar:
+            var prices = GetPricesCheckingPythonWrapper(asset, order.Direction);
+            var pricesEndTime = prices.EndTime.ConvertToUtc(asset.Exchange.TimeZone);
+
+            // do not fill on stale data
+            if (pricesEndTime <= order.Time) return fill;
+
+            //Calculate the model slippage: e.g. 0.01c
+            var slip = asset.SlippageModel.GetSlippageApproximation(asset, order);
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/FillModel.cs:340]: >             //Calculate the model slippage: e.g. 0.01c
>            var slip = asset.SlippageModel.GetSlippageApproximation(asset, order);
            
Think we should remove those 👍

File: Common/Orders/Fills/EquityFillModel.cs [hunk 14]
```diff
@@ -30,18 +30,18 @@ namespace QuantConnect.Orders.Fills
...
         /// This is required due to a limitation in PythonNet to resolved overriden methods.
         /// When Python calls a C# method that calls a method that's overriden in python it won't
         /// run the python implementation unless the call is performed through python too.
         /// </summary>
         protected FillModelPythonWrapper PythonWrapper;
 
+        /// <summary>
+        /// The parameters instance to be used by the different XxxxFill() implementations
+        /// </summary>
+        protected FillModelParameters Parameters { get; set; }
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:43]: Let's avoid changes like these 👍

File: Common/Orders/Fills/EquityFillModel.cs [hunk 15]
```diff
@@ -98,9 +103,122 @@ public virtual Fill Fill(FillModelParameters parameters)
...
+            }
+
+            // Get the range of prices in the last bar:
+            var high = 0m;
+            var low = 0m;
+            var open = 0m;
+            var endTimeUtc = DateTime.MinValue;
+
+            var subscribedTypes = GetSubscribedTypes(asset);
+            var slip = asset.SlippageModel.GetSlippageApproximation(asset, order);
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:148]: There should be no slippage?

File: Common/Orders/Fills/EquityFillModel.cs [hunk 16]
```diff
@@ -98,9 +103,122 @@ public virtual Fill Fill(FillModelParameters parameters)
...
+
+                    break;
+
+                case OrderDirection.Sell:
+                    //-> 1.2 Sell: If Price above Trigger, Sell:
+                    if (high >= order.TriggerPrice || order.TriggerTouched)
+                    {
+                        order.TriggerTouched = true;
+
+                        if (high >= order.LimitPrice)
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:205]: Cool! So we use trade information to determine if the order is triggered, first if. This second if, should already be using quote information to determine if the limit order is filled. There should be no `slip`. 
If `bidOpen >= order.LimitPrice` think we can fill at `bidOpen`, else `Math.Max(GetCurrentBidPrice(asset, out endTimeUtc), order.LimitPrice)`. Not `ask`, `ask` is for the `OrderDirection.Buy`.

Some references: https://github.com/QuantConnect/Lean/issues/4567 IB SL https://www1.interactivebrokers.com/en/index.php?f=608 IB LIT https://www1.interactivebrokers.com/en/index.php?f=592

File: Common/Orders/Fills/FillModel.cs [hunk 17]
```diff
@@ -286,6 +291,88 @@ public virtual OrderEvent StopLimitFill(Security asset, StopLimitOrder order)
...
+                    if (prices.High > order.TriggerPrice || order.TriggerTouched)
+                    {
+                        order.TriggerTouched = true;
+
+                        //-> 1.1 Limit surpassed: Sell.
+                        if (prices.Current > order.LimitPrice)
+                        {
+                            fill.Status = OrderStatus.Filled;
+                            // Assuming worse case scenario fill - fill at lowest of the limit & asset price.
+                            fill.FillPrice = Math.Min(order.LimitPrice, prices.Low);
```
[the remark this answers is not in the document]
Comment (aarjaneiro) [reply Common/Orders/Fills/FillModel.cs]: Should make sense to have this changed to prices.Current, right?

File: Algorithm.CSharp/LimitIfTouchedRegressionAlgorithm.cs [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+            if (data.ContainsKey("SPY"))
+            {
+                if (Time.Second == 0 && Time.Minute == 0 && Transactions.GetOpenOrders().Count < 1)
+                {
+                    var goLong = Time < StartDate.AddDays(2);
+                    var negative = goLong ? 1 : -1;
+                    request = new SubmitOrderRequest(OrderType.LimitIfTouched, SecurityType.Equity, "SPY",
+                        negative * 10, 0, data["SPY"].Price - (decimal) 0.5 * negative,
+                        data["SPY"].Price - (decimal) negative, UtcTime, $"LIT - {UtcTime}");
+                    Transactions.AddOrder(request);
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/LimitIfTouchedRegressionAlgorithm.cs:60]: Curious why this approach versus calling `LimitIfTouchedOrder()`? Let's please add an order update of the LIT order asserting it worked, simple just the quantity

File: Common/Orders/Fills/EquityFillModel.cs [hunk 19]
```diff
@@ -30,17 +30,17 @@ namespace QuantConnect.Orders.Fills
     /// </summary>
     public class EquityFillModel : IFillModel
     {
-        /// <summary>
-        /// The parameters instance to be used by the different XxxxFill() implementations
-        /// </summary>
-        protected FillModelParameters Parameters { get; set; }
-
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:37]: Let's revert these changes 👍

File: Common/Orders/Fills/EquityFillModel.cs (hunk 16 above)
Comment (Martin-Molinero) [reply Common/Orders/Fills/EquityFillModel.cs:212]: This wasn't addressed @aarjaneiro? There should be a failing test case for this, let's add a unit test too 👍

File: Common/Orders/Fills/EquityFillModel.cs [hunk 20]
```diff
@@ -143,7 +268,8 @@ public virtual OrderEvent MarketFill(Security asset, MarketOrder order)
             // if the order is filled on stale (fill-forward) data, set a warning message on the order event
             if (endTimeUtc.Add(Parameters.StalePriceTimeSpan) < order.Time)
             {
-                fill.Message = $"Warning: fill at stale price ({endTime.ToStringInvariant()} {asset.Exchange.TimeZone})";
+                fill.Message =
+                    $"Warning: fill at stale price ({endTime.ToStringInvariant()} {asset.Exchange.TimeZone})";
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:272]: Let's revert these format changes/white spaces

File: Common/Orders/Fills/EquityFillModel.cs [hunk 21]
```diff
@@ -612,87 +829,8 @@ private Prices GetPricesCheckingPythonWrapper(Security asset, OrderDirection dir
...
-            {
-                // if we're not open at the current time exactly, check the bar size, this handle large sized bars (hours/days)
-                var currentBar = asset.GetLastData();
-                if (asset.LocalTime.Date != currentBar.EndTime.Date
-                    || !asset.Exchange.IsOpenDuringBar(currentBar.Time, currentBar.EndTime, isExtendedMarketHours))
-                {
-                    return false;
-                }
-            }
-            return true;
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:695]: Please revert all these changes 👍

File: Common/Orders/Fills/FillModel.cs [hunk 22]
```diff
@@ -286,6 +301,86 @@ public virtual OrderEvent StopLimitFill(Security asset, StopLimitOrder order)
...
+            if (!IsExchangeOpen(asset,
+                Parameters.ConfigProvider
+                    .GetSubscriptionDataConfigs(asset.Symbol)
+                    .IsExtendedMarketHours()))
+            {
+                return fill;
+            }
+            
+            //Get the range of prices in the last bar:
+            var prices = GetPricesCheckingPythonWrapper(asset, order.Direction);
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/FillModel.cs:343]: Shouldn't use `prices` but trades and quotes, same as `EquityFillModel`

File: Common/Orders/Fills/FillModel.cs [hunk 23]
```diff
@@ -286,6 +301,86 @@ public virtual OrderEvent StopLimitFill(Security asset, StopLimitOrder order)
...
+                    if (prices.Current > order.TriggerPrice || order.TriggerTouched)
+                    {
+                        order.TriggerTouched = true;
+
+                        //-> 1.1 Limit surpassed: Sell.
+                        if (prices.Current > order.LimitPrice)
+                        {
+                            fill.Status = OrderStatus.Filled;
+                            // Assuming worse case scenario fill - fill at lowest of the limit & asset price.
+                            fill.FillPrice = order.LimitPrice;
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/FillModel.cs:361]: Nice 👍
Comment is off `// Assuming worse case scenario fill - fill at lowest of the limit & asset price.`

File: Common/Orders/OrderEvent.cs [hunk 24]
```diff
@@ -278,6 +299,10 @@ public string ShortToString()
             {
                 message += Invariant($" SP:{StopPrice.Value.SmartRounding()}");
             }
+            if (TriggerPrice.HasValue)
+            {
+                message += Invariant($" TP:{TriggerPrice.Value.SmartRounding()}");
```
Comment (Martin-Molinero) [inline Common/Orders/OrderEvent.cs:304]: Thanks!

File: Tests/Common/Orders/Fills/ImmediateFillModelTests.cs [hunk 25]
```diff
@@ -349,7 +349,91 @@ public void PerformsStopMarketFillSell()
...
+
+            fill = model.Fill(new FillModelParameters(
+                security,
+                order,
+                new MockSubscriptionDataConfigProvider(config),
+                Time.OneHour)).OrderEvent;
+
+            // this fills worst case scenario, so it's min of asset/stop price
+            Assert.AreEqual(order.Quantity, fill.FillQuantity);
+            Assert.AreEqual(Math.Min(security.Price, order.LimitPrice), fill.FillPrice);
```
Comment (Martin-Molinero) [inline Tests/Common/Orders/Fills/ImmediateFillModelTests.cs:433]: `FillModel` doesn't at worst case scenario anymore? Same for `PerformsLimitIfTouchedFillBuy`

File: Algorithm.CSharp/LimitIfTouchedRegressionAlgorithm.cs (hunk 18 above)
Comment (aarjaneiro) [reply Algorithm.CSharp/LimitIfTouchedRegressionAlgorithm.cs:60]: Was to display that order-interactions were possible but agree it would make more sense to utilize it by changing the order quantity!

File: Common/Orders/Fills/FillModel.cs (hunk 22 above)
Comment (aarjaneiro) [reply Common/Orders/Fills/FillModel.cs:343]: I imagined that would be preferrable, though ```private Prices GetPricesCheckingPythonWrapper(Security asset, OrderDirection direction)``` is used in fillmodel versus equityfillmodel's ```GetSubscribedTypes```. Should an analogue of ```GetSubscribedTypes```, say ```GetQuotesCheckingPythonWrapper``` be added in this PR for allowing use of quotebars?

File: Common/Orders/Fills/EquityFillModel.cs (hunk 16 above)
Comment (aarjaneiro) [reply Common/Orders/Fills/EquityFillModel.cs:212]: Apologies -- I thought I changed open -> bidopen!

File: Common/Orders/Fills/FillModel.cs [hunk 26]
```diff
@@ -286,6 +303,121 @@ public virtual OrderEvent StopLimitFill(Security asset, StopLimitOrder order)
...
+
+            switch (order.Direction)
+            {
+                case OrderDirection.Sell:
+                    if (tradeHigh >= order.TriggerPrice || order.TriggerTouched)
+                    {
+                        order.TriggerTouched = true;
+
+                        //-> 1.1 Limit surpassed: Sell.
+                        if (tradeHigh >= order.LimitPrice)
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/FillModel.cs:394]: `tradeHigh >= order.LimitPrice` This should be using quotes? Please add a failing unit test for this issue

File: Common/Orders/Fills/EquityFillModel.cs (hunk 19 above)
Comment (aarjaneiro) [reply Common/Orders/Fills/EquityFillModel.cs:37]: Seems to be reverted although github isnt showing such? https://github.com/aarjaneiro/Lean/blob/930d818d5878e57862fd073f464d6c2bbca25dd9/Common/Orders/Fills/FillModel.cs#L33

File: Common/Orders/Fills/EquityFillModel.cs [hunk 27]
```diff
@@ -101,6 +106,121 @@ public virtual Fill Fill(FillModelParameters parameters)
...
+                    tradeLow = trade.Price;
+                    bidOpen = quote.BidPrice;
+                    askOpen = quote.AskPrice;
+                    endTimeUtc = trade.EndTime.ConvertToUtc(asset.Exchange.TimeZone);
+                }
+            }
+            else if (subscribedTypes.Contains(typeof(TradeBar)))
+            {
+                var tradeBar = asset.Cache.GetData<TradeBar>();
+                var quoteBar = asset.Cache.GetData<QuoteBar>();
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:165]: `QuoteBar` is another subscription type really, if at 162 doesn't apply

File: Common/Orders/Fills/EquityFillModel.cs [hunk 28]
```diff
@@ -101,6 +106,121 @@ public virtual Fill Fill(FillModelParameters parameters)
...
+
+                if (tradeBar != null && quoteBar != null)
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:167]: `tradeBar != null && quoteBar != null` these aren't related, should be independent

File: Common/Orders/Fills/EquityFillModel.cs [hunk 29]
```diff
@@ -101,6 +106,121 @@ public virtual Fill Fill(FillModelParameters parameters)
...
+            //Check if the limit if touched order was filled:
+            switch (order.Direction)
+            {
+                case OrderDirection.Buy:
+                    //-> 1.2 Buy: If Price below Trigger, Buy:
+                    if (tradeLow <= order.TriggerPrice || order.TriggerTouched)
+                    {
+                        order.TriggerTouched = true;
+
+                        if (tradeLow <= order.LimitPrice)
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:189]: `tradeLow <= order.LimitPrice` I believe it should use quote information here

File: Common/Orders/Fills/EquityFillModel.cs [hunk 30]
```diff
@@ -462,7 +582,7 @@ public virtual OrderEvent MarketOnCloseFill(Security asset, MarketOnCloseOrder o
 
             return fill;
         }
-
+
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:585]: nit - another tab space

File: Algorithm.CSharp/LimitIfTouchedRegressionAlgorithm.cs [hunk 31]
```diff
@@ -0,0 +1,176 @@
...
+
+namespace QuantConnect.Algorithm.CSharp
+{
+    /// <summary>
+    /// Basic algorithm demonstrating how to place LimitIfTouched orders.
+    /// </summary>
+    /// <meta name="tag" content="trading and orders" />
+    /// <meta name="tag" content="placing orders" />`
+    /// <meta name="tag" content="limit if touched order"/>
+    public class LimitIfTouchedRegressionAlgorithm : QCAlgorithm, IRegressionAlgorithmDefinition
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/LimitIfTouchedRegressionAlgorithm.cs:30]: Regression algorithm is failing

File: Common/Orders/Fills/EquityFillModel.cs [hunk 32]
```diff
@@ -101,6 +107,109 @@ public virtual Fill Fill(FillModelParameters parameters)
...
+                    break;
+
+                case OrderDirection.Sell:
+                    //-> 1.2 Sell: If Price above Trigger, Sell:
+                    if (tradeHigh >= order.TriggerPrice || order.TriggerTouched)
+                    {
+                        order.TriggerTouched = true;
+                        var bidCurrent = GetBidPrice(asset, out endTimeUtc);
+
+                        if (bidCurrent >= order.LimitPrice)
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:200]: Here we can't know the order of things happening in a bar: if the LIT order was triggered, using trades, before the quote low, so we use **current** quote information instead. In practice, it might leave trades out of happening. This is following the existing behavior of `StopLimitFill`. Let's please add this as a comment in the description of this method and the `FillModel` version too 👍

**Review by Martin-Molinero — APPROVED** [review]
Looks good, minor comment request and regression algorithm is failing

**Review by Martin-Molinero — APPROVED** [review]

**Review by Martin-Molinero — APPROVED** [review]
Thanks @aarjaneiro!
